Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
+		if (m_OwnerMO->OnMOHit(intersectedMO)) {
+			return false;
+		}
+		if (intersectedMO->OnMOHit(m_OwnerMO->GetRootParent())) {
+			return false;
+		}
+
+		std::list<Atom *> intersectingAtoms;
+
+		// Restart and go through all Atoms to find all intersecting the specific intersected MO
+		for (Atom *atom : m_Atoms) {
+			atomPos = atom->GetCurrentPos();
+			if (g_SceneMan.GetMOIDPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY()) == currentMOID) {
+				// Add atom to list of intersecting ones
+				intersectingAtoms.push_back(atom);
+			}
+		}
+
+		Vector exitDirection = Vector();
+		Vector atomExitVector = Vector();
+		Vector totalExitVector = Vector();
+
+		// Go through all intersecting Atoms and find their average inverse normal
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			exitDirection += m_OwnerMO->RotateOffset(intersectingAtom->GetNormal());
+		}
+
+		// We don't have a direction to go, so quit
+		// TODO: Maybe use previous position to create an exit direction instead then?
+		if (exitDirection.IsZero()) {
+			return false;
+		}
+
+		// Invert and set appropriate length
+		exitDirection = -exitDirection;
+		exitDirection.SetMagnitude(m_OwnerMO->GetDiameter());
+
+		Vector clearPos = Vector();
+
+		// See which of the intersecting Atoms has the longest to travel along the exit direction before it clears
+		float longestDistance = 0.0F;
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			atomPos = intersectingAtom->GetCurrentPos();
+			if (g_SceneMan.CastFindMORay(atomPos, exitDirection, g_NoMOID, clearPos, 0, true, 0)) {
+				// Determine the longest clearing distance so far
+				atomExitVector = clearPos - atomPos.GetFloored();
+				if (atomExitVector.GetMagnitude() > longestDistance) {
+					// We found the Atom with the longest to travel along the exit direction to clear, so that's the distance to move the whole object to clear all its Atoms.
+					longestDistance = atomExitVector.GetMagnitude();
+					totalExitVector = atomExitVector;
+				}
+			}
+		}
+
+		// The final exit movement vectors for this' Owner MO and the MO intersected by this
+		Vector thisExit;
+		Vector intersectedExit;
+
+		// If the other object is pinned, then only move this. Otherwise, apply the object exit vector to both this owner object, and the one it intersected, proportional to their masses.
+		if (intersectedMO->GetPinStrength() > 0.0F) {
+			thisExit = totalExitVector;
+		} else {
+			float massA = m_OwnerMO->GetMass();
+			float massB = intersectedMO->GetMass();
+			float invMassA = 1.0F / massA;
+			float invMassB = 1.0F / massB;
+			float normMassA = invMassA / (invMassA + invMassB);
+			float normMassB = invMassB / (invMassA + invMassB);
+
+			// If the intersected is much larger than this' MO, then only move this. Otherwise, apply the movements to both this and the intersected MO's, proportional to their respective masses.
+			if (normMassB < 0.33F) {
```

Comment:
Seems a little odd this doesn't go both ways, i.e. if this object is a lot heavier than the hit one this still gets moved. Granted, since this is all affected by mass it won't move much, so I guess it's fine, but I wonder if it'd be better if this were consistent.

If you were to change this and add the comment I mentioned above, make sure said comment mentions this idea too.